_drafts/firmware-code-size-2.md

```diff
@@ -0,0 +1,303 @@
+
+Since the compiler operates on one file at a time, it does not have enough context to decide whether a function is dead code or not.  Consider a library which exposes a function to add numbers in an array: since the files in the library do not call that function, the compiler could think it is not used when other parts of our program rely on it.
+
+The linker on the other hand has visibility into our whole program, so this is where dead code could be identified and removed. Unfortunately, linkers do not perform optimizations by default.
+
+To enable dead code optimization on GCC, you need two things: the compiler needs to split each function into its own linker section so the linker knows where each function is, and the linker needs to add an optimization pass to remove sections that are not called by anything.
+
+This is achieved with the `-ffunction-sections`  compile-time flag, and the `-gc-sections` link-time flag. A similar process can take place with dead data and the `-fdata-sections` flag.
+
+In a Makefile, it looks like this:
+```makefile
+CFLAGS += -ffunction-sections -fdata-sections
+LDFLAGS += -Wl,--gc-sections
+```
+
+ChibiOS sets those when the `USE_LINK_GC` variable is set, so we call:
+```bash
+$ make USE_LINK_GC=yes USE_LTO=no
+```
+
+And get:
+```bash
+Linking build/ch.elf
+Creating build/ch.hex
+Creating build/ch.bin
+Creating build/ch.dmp
+
+   text	   data	    bss	    dec	    hex	filename
+  54828	    228	  98072	 153128	  25628	build/ch.elf
+Creating build/ch.list
+```
+
+So dead code elimination yields close to another 10KiB!
+
+### Preventing some symbols from being garbage collected
+In some cases, you may want some code that is not explicitly called by anything else to remain in your program. One such example is interrupt handlers: while it may look to the linker like they are not called, they the hardware will jump to those addresses and expect the code to be there.
+
+There are two ways to make sure a symbol isn’t garbage collected: with the `used` attribute, or with the `KEEP` linker script command.
+
+Here’s how the `used` attribute is used:
+```c
+void foo(void) __attribute__((used))
+{
+   // ...
+}
+```
+
+For interrupts, the typical solution is to put your vector table in a section called `vectors` and mark it in the linker script:
+
+```
+    .text :
+    {
+        KEEP(*(.vectors .vectors.*))
+  	      [...]
+    } > rom
+```
+
+
+## Link-time optimization
+While linkers do not traditionally do much optimizing, this has started to change. Nowadays, all the major compilers offer Link Time Optimization or LTO. LTO enables optimizations that are not available to the compiler as they touch multiple compilation units at once.
+
+LTO comes at the cost of compilation time. Even for simple programs the difference in build time will be noticeable, some of your colleagues may complain about it!
```

Comment:
Maybe

LTO comes with two major draw backs:
* compilation time ...
* debugability ...

fwiw, a few other drawbacks I can think of:

* deeper stack depths due to aggressive inlining (dealt with a couple stack overflows due to this)
* still all-or-nothing afaik, you cannot LTO certain units. So if you are using a topology where memory placement of certain items is important, its hard to accomplish